### PR TITLE
Add a new prop in message thread to allow customization of date time format

### PR DESCRIPTION
**Pull request by carocao-msft**
# What
Add a new prop in message thread to allow customization of date time format
Add a new prop in localization provider to allow contoso to pass in date time formatting function on the composite level
Add examples in storybook describing how to customize date time format
Add test for message thread to test both composite and component level customization

# Why

https://github.com/Azure/communication-ui-library/issues/1862

Screen Shot 2022-05-16 at 8 36 34 PM

Screen Shot 2022-05-06 at 9 33 25 AM

Screen Shot 2022-05-06 at 9 33 32 AM

### CODE REVIEW

File: packages/react-components/src/components/ChatMessage/ChatMessageComponent.tsx [hunk 1]
```diff
@@ -52,6 +52,11 @@ type ChatMessageComponentProps = {
    * @param userId - user Id
    */
   onRenderAvatar?: OnRenderAvatarCallback;
+  /**
+   * Optional function to provide customized date format.
+   *
+   */
+  customizedFormatTimestampForChatMessage?: (messageDate: Date) => string;
```
Comment (alkwa-msft) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponent.tsx:59]: should we just have 

dateTimeFormat as the name of the prop that is optional?
If its set we use it and if we don't have it or its invalid we use the default behavior?

File: packages/react-components/src/components/MessageThread.tsx [hunk 2]
```diff
@@ -601,6 +601,12 @@ export type MessageThreadProps = {
    * this function will be called with the data inside `fileSharingMetadata` key.
    */
   fileDownloadHandler?: FileDownloadHandler;
+
+  /**
+   * Optional function to provide customized date format.
+   *
+   */
+  customizedFormatTimestampForChatMessage?: (messageDate: Date) => string;
```
Comment (alkwa-msft) [inline packages/react-components/src/components/MessageThread.tsx:609]: can just call this prop "dateTimeFormat"

File: packages/react-components/src/components/MessageThread.tsx [hunk 3]
```diff
@@ -966,6 +973,7 @@ export const MessageThread = (props: MessageThreadProps): JSX.Element => {
             showMessageStatus={showMessageStatus}
             messageStatus={messageProps.message.status}
             onActionButtonClick={onActionButtonClickMemo}
+            customizedFormatTimestampForChatMessage={customizedFormatTimestampForChatMessage}
```
Comment (alkwa-msft) [inline packages/react-components/src/components/MessageThread.tsx:976]: should it be an optionalProp called

dateTimeFormat? would we use it for anything other than chat messages?

Comment (carocao-msft) [reply packages/react-components/src/components/MessageThread.tsx]: updated

File: packages/react-components/src/components/MessageThread.tsx [hunk 4]
```diff
@@ -966,6 +973,7 @@ export const MessageThread = (props: MessageThreadProps): JSX.Element => {
             showMessageStatus={showMessageStatus}
             messageStatus={messageProps.message.status}
             onActionButtonClick={onActionButtonClickMemo}
+            dateTimeFormat={dateTimeFormat}
```
Comment (prabhjot-msft) [inline packages/react-components/src/components/MessageThread.tsx:976]: As `dateTimeFormat` is an optional prop, don't we need to add a check if prop is available or not.

**Review by prabhjot-msft — APPROVED** [review]
minor comment, otherwise looks good to me!

File: packages/react-components/src/components/ChatMessage/ChatMessageComponent.tsx [hunk 5]
```diff
@@ -52,6 +52,11 @@ type ChatMessageComponentProps = {
    * @param userId - user Id
    */
   onRenderAvatar?: OnRenderAvatarCallback;
```
Comment (PorterNan) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponent.tsx:54]: I am a little bit curious about what layer customers are asking for the dateTimeFormat? Is it component layer or composite layer?

File: packages/communication-react/stable-review/communication-react.api.md [hunk 6]
```diff
@@ -1805,6 +1805,7 @@ export type MessageThreadProps = {
     onSendMessage?: (messageId: string) => Promise<void>;
     disableEditing?: boolean;
     strings?: Partial<MessageThreadStrings>;
+    dateTimeFormat?: (messageDate: Date) => string;
```
Comment (PorterNan) [inline packages/communication-react/stable-review/communication-react.api.md:1808]: I am thinking of a more generic way of this function - what if customer wants to further customized their timeStamp? What if they want to hover over time stamp and show some special popup? what if they want to print a img in timestamp?
There could be another option like onRenderTimeStamp?: (messageDate: Date) => React.Component to open all these possibilities (especially in component layer, then we might be able to set more restriction in composite level)

I would like to involve @anjulgarg to this discussion, cause he's the only expert on custom data injection in our team

File: packages/communication-react/review/communication-react.api.md [hunk 7]
```diff
@@ -1942,6 +1942,7 @@ export type MessageThreadProps = {
     disableEditing?: boolean;
     strings?: Partial<MessageThreadStrings>;
     fileDownloadHandler?: FileDownloadHandler;
+    dateTimeFormat?: (messageDate: Date) => string;
```
Comment (anjulgarg) [inline packages/communication-react/review/communication-react.api.md:1945]: This prop name is misleading to me.
When I think of `date time format` I think of the string template used for formatting a date.
For example, `YY-MM-DD` etc.
However, this callback is supposed to return the formatted datetime string which is confusing to me.

Comment (anjulgarg) [inline packages/communication-react/review/communication-react.api.md:1945]: The callback is named `dateTimeFormat` but the input parameter is `messageDate`
A better name would be `messageDateTime` as it is inline with the callback name.

Comment (anjulgarg) [reply packages/communication-react/review/communication-react.api.md:1945]: OR changing the callback name to `dateFormat` is another option

Comment (anjulgarg) [inline packages/communication-react/review/communication-react.api.md:1945]: How about we go 1 step further?
Since Javascript doesn't have native date formatting using `YYYY, MM, DD` etc.
We can use a popular library like momentjs, datejs behind the scenes and accept a date time format making the implementation standardized and inline with other programming languages

Comment (anjulgarg) [reply packages/communication-react/review/communication-react.api.md:1945]: OR use the native platform supported formatting features like
image

Comment (anjulgarg) [reply packages/communication-react/review/communication-react.api.md:1945]: Here's an article to help you out
https://dockyard.com/blog/2020/02/14/you-probably-don-t-need-moment-js-anymore

File: packages/react-components/src/components/MessageThread.tsx (hunk 2 above)
Comment (alkwa-msft) [reply packages/react-components/src/components/MessageThread.tsx]: this is really a callback that gives a dateTime and wants to return a string

I am thinking something like on__________ would be a better naming convention? 
onRenderDateTime doesn't seem quite right since it kind of implies returning a JSX.Element based on our other onRenders?

I think it might be good to have the same named prop in ComponentLocale and everywhere else in the library?
@PorterNan  thoughts?

File: packages/react-components/src/components/ChatMessage/ChatMessageComponent.tsx (hunk 5 above)
Comment (alkwa-msft) [reply packages/react-components/src/components/ChatMessage/ChatMessageComponent.tsx:61]: It should really be in both. The components need to handle it and we want to expose it at the composite level.

@PorterNan  can we confirm this will be exposed through the JS bundle as well?

File: packages/communication-react/stable-review/communication-react.api.md (hunk 6 above)
Comment (alkwa-msft) [reply packages/communication-react/stable-review/communication-react.api.md]: Lets just only focus on strings for now and we can look into a modification at the component level if devs want to extremely customize this area

**Comment by alkwa-msft** [comment]
I don't think we want to to expose this API under we have reviewed this with ARB. Lets only expose this at the beta level for now if we want to.

File: packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx [hunk 8]
```diff
@@ -145,11 +155,18 @@ const MessageBubble = (props: ChatMessageComponentAsMessageBubbleProps): JSX.Ele
...
           timestamp={
             <Text data-ui-id={ids.messageTimestamp}>
-              {message.createdOn
-                ? showDate
-                  ? formatTimestampForChatMessage(message.createdOn, new Date(), strings)
-                  : formatTimeForChatMessage(message.createdOn)
-                : undefined}
+              {
+                // messageDateTime overwrites messageDateTimeLocale overwrites default
+                message.createdOn
```
[the remark this answers is not in the document]
Comment (carocao-msft) [reply packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx]: Updated the code and made the logic more clear, also added conditional compile. messageDateTimeLocale here comes from const { messageDateTimeLocale } = useLocale(); and the messageDateTime comes from prop passed to message thread component

File: packages/react-components/src/components/MessageThread.test.tsx [hunk 9]
```diff
@@ -38,3 +61,54 @@ describe('Message date should be formatted correctly', () => {
     expect(component.text()).toContain(testLocale.strings.messageThread.yesterday);
   });
 });
+
+describe('Message date should be customized by function passed through locale', () => {
+  test('Message date should be localized to "24 hours ago"', async () => {
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/MessageThread.test.tsx:66]: `/* @conditional-compile-remove fancy-feature-name */`

**Comment by prprabhu-ms** [comment]
Removed a couple assignees to create clearer responsibility for moving the review along. Feel free to modify again. As a thumb rule, set 2-3 assignees. Any more and it dilutes responsibility for who needs to review proactively.

File: change/@internal-react-components-25d9c1f2-2a56-460a-9b8b-7a11ae28996f.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "prerelease",
+  "comment": "Added new prop to chat message to allow custom date/time format",
```
Comment (anjulgarg) [inline change/@internal-react-components-25d9c1f2-2a56-460a-9b8b-7a11ae28996f.json:3]: Adding prop name here would be helpful when changelog is created :)

File: packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx [hunk 11]
```diff
@@ -76,9 +87,28 @@ const MessageBubble = (props: ChatMessageComponentAsMessageBubbleProps): JSX.Ele
     onRenderAvatar,
     showMessageStatus,
     messageStatus,
-    fileDownloadHandler
+    fileDownloadHandler,
+    /* @conditional-compile-remove(dateTimeCustomization) */
+    onDisplayDateTimeString
   } = props;
 
+  // onDisplayDateTimeString overwrites onDisplayDateTimeString overwrites default
```
Comment (anjulgarg) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx:95]: onDisplayDateTimeString overwrites default

File: packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx [hunk 12]
```diff
@@ -76,9 +87,28 @@ const MessageBubble = (props: ChatMessageComponentAsMessageBubbleProps): JSX.Ele
...
+  let formattedTimestamp: string | undefined = undefined;
+  if (message.createdOn) {
+    // default
+    formattedTimestamp = showDate
+      ? formatTimestampForChatMessage(message.createdOn, new Date(), strings)
+      : formatTimeForChatMessage(message.createdOn);
+
+    /* @conditional-compile-remove(dateTimeCustomization) */
+    if (onDisplayDateTimeString) {
```
Comment (anjulgarg) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx:104]: Should we use onDisplayDateTimeString if showDate is false?

Comment (carocao-msft) [reply packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx:104]: when ShowDate is false, timestamps are still displayed but displayed without dates, only time. So show date is a way to customize default date time format. onDisplayDateTimeString should overwrite showdate

File: packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx (hunk 11 above)
Comment (carocao-msft) [reply packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx]: Updated

File: change/@internal-react-components-25d9c1f2-2a56-460a-9b8b-7a11ae28996f.json (hunk 10 above)
Comment (carocao-msft) [reply change/@internal-react-components-25d9c1f2-2a56-460a-9b8b-7a11ae28996f.json]: updated

File: packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx [hunk 13]
```diff
@@ -76,9 +87,28 @@ const MessageBubble = (props: ChatMessageComponentAsMessageBubbleProps): JSX.Ele
...
+      formattedTimestamp = onDisplayDateTimeString(message.createdOn);
+    } else if (onDisplayDateTimeStringLocale) {
+      /* @conditional-compile-remove(dateTimeCustomization) */
```
Comment (anjulgarg) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx:107]: Is this conditional compilation flag unnecessary here?
the one above the if block should remove the entire conditional block

File: packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx [hunk 14]
```diff
@@ -76,9 +87,28 @@ const MessageBubble = (props: ChatMessageComponentAsMessageBubbleProps): JSX.Ele
     onRenderAvatar,
     showMessageStatus,
     messageStatus,
-    fileDownloadHandler
+    fileDownloadHandler,
+    /* @conditional-compile-remove(dateTimeCustomization) */
+    onDisplayDateTimeString
   } = props;
 
+  // onDisplayDateTimeString from messagethread overwrites onDisplayDateTimeString from locale overwrites default
```
Comment (anjulgarg) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx:95]: This logic is hard to follow. Can we simplify it by doing something like this?
image

**Review by anjulgarg — APPROVED** [review]

**Comment by anjulgarg** [comment]
Although I have approved, I would recommend getting another review from @prprabhu-ms

File: common/config/babel/.babelrc.js [hunk 15]
```diff
@@ -23,6 +23,8 @@ process.env['COMMUNICATION_REACT_FLAVOR'] === 'stable' &&
         'file-sharing',
         // Adhoc calls to a Teams user.
         'teams-adhoc-call',
+        // props to allow Contoso to overwrite timestamp format for chat messages, one in locale and one in message thread component
+        'dateTimeCustomization',
```
Comment (prprabhu-ms) [inline common/config/babel/.babelrc.js:27]: Can we rename this to match the pattern we've used in other features: `date-time-customization`.

File: packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx [hunk 16]
```diff
@@ -79,6 +102,16 @@ const MessageBubble = (props: ChatMessageComponentAsMessageBubbleProps): JSX.Ele
     fileDownloadHandler
   } = props;
 
+  const defaultTimeStamp = message.createdOn
+    ? generateDefaultTimestamp(message.createdOn, showDate, strings)
+    : undefined;
+
+  // onDisplayDateTimeString from props overwrite onDisplayDateTimeString from locale
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx:109]: nit: Move comment to the implementatino of `generatCustomizedTimestamp`

File: packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx [hunk 17]
```diff
@@ -79,6 +102,16 @@ const MessageBubble = (props: ChatMessageComponentAsMessageBubbleProps): JSX.Ele
     fileDownloadHandler
   } = props;
 
+  const defaultTimeStamp = message.createdOn
+    ? generateDefaultTimestamp(message.createdOn, showDate, strings)
+    : undefined;
+
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx:108]: nitty, nitty, nit: I'd drop the newlines between these variables (they're all related)

File: packages/react-components/src/components/ChatMessage/ChatMessageComponent.tsx [hunk 18]
```diff
@@ -59,6 +59,13 @@ type ChatMessageComponentProps = {
...
+
+  /* @conditional-compile-remove(dateTimeCustomization) */
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponent.tsx:63]: `ChatMessageComponentProps` is not exported from this module (and from the public API).

So there is no need for any conditional compilation here. Please add it unconditionally (this also simplifies your implementation).

Only public API needs to be added conditionally.

File: packages/react-components/src/components/utils/Datetime.ts [hunk 19]
```diff
@@ -114,3 +114,18 @@ const dayToDayName = (day: number, dateStrings: MessageThreadStrings): string =>
       throw new Error(`Invalid day [${day}] passed`);
   }
 };
+
+/**
+ * @private
+ */
+export const generateDefaultTimestamp = (
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/utils/Datetime.ts:121]: nit: I'd keep this helper function in `ChatMessageComponent` related files. This is not generally applicable to anything beyond a message (The `showDate` prop etc are very specific), so no value in moving this to a generic utils file. Instead, moving it here just increases the cross-references in the code-base.

General rule: Keep the modules as self contained as possible. Move generic functionality out to common modules / other modules when it is truly generic (has at least 2 (potential) use-cases in the codebase)

File: packages/storybook/stories/ChatComposite/snippets/CustomDateTimeFormat.snippet.tsx [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+    {
+      endpoint: props.endpointUrl,
+      userId,
+      // Data model injection: The display name for the local user comes from Contoso's data model.
+      displayName: props.displayName,
+      credential,
+      threadId: props.threadId
+    },
+    // Custom behavior: Capitalize all messages before sending them.
+    capitalizeMessageBeforeSend
```
Comment (prprabhu-ms) [inline packages/storybook/stories/ChatComposite/snippets/CustomDateTimeFormat.snippet.tsx:42]: Copy-pasta example. Please remove this custom behavior from the example.

File: packages/storybook/stories/ChatComposite/CustomDateTimeFormat.stories.tsx [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+      } else {
+        setContainerProps(undefined);
+      }
+    };
+    fetchToken();
+  }, [args.userId, args.token, args.botId, args.botToken, args.endpointUrl, args.displayName]);
+
+  const strings = compositeLocale(locale)?.component.strings ?? COMPONENT_LOCALE_EN_US.strings;
+  const compositeStrings = compositeLocale(locale)?.strings ?? COMPOSITE_LOCALE_EN_US.strings;
+  const onDisplayDateTimeStringLocale =
```
Comment (prprabhu-ms) [inline packages/storybook/stories/ChatComposite/CustomDateTimeFormat.stories.tsx:94]: Why this fallback? We don't set the datetime format via the storybook locale at all. So let's directly assign the local function.

File: packages/storybook/stories/ChatComposite/CustomDateTimeFormat.stories.tsx [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+  'Have fun!'
+];
+
+const CustomDateTimeFormatStory = (args, context): JSX.Element => {
+  const {
+    globals: { locale }
+  } = context;
+  const [containerProps, setContainerProps] = useState<ChatCompositeSetupProps>();
+
+  const onDisplayDateTimeString = (messageDate: Date): string => {
```
Comment (prprabhu-ms) [inline packages/storybook/stories/ChatComposite/CustomDateTimeFormat.stories.tsx:33]: Please move this function to file-level private function. We should avoid creating functions nested inside other functions because they reduce readability. You need to do this if you are capturing some local variable, even there it's better to create a helper function with most of the logic and call that from your closure defined here.

File: packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx [hunk 23]
```diff
@@ -57,20 +57,33 @@ type ChatMessageComponentAsMessageBubbleProps = {
    */
   onRenderAvatar?: OnRenderAvatarCallback;
 
-  /* @conditional-compile-remove(dateTimeCustomization) */
+  /* @conditional-compile-remove(date-time-customization) */
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/ChatMessage/ChatMessageComponentAsMessageBubble.tsx:60]: This can still be added unconditionally (not part of public API).

I'll make your code simpler by needing less conditional compilation.

**Review by prprabhu-ms — APPROVED** [review]
